src/vma/dev/qp_mgr_eth_mlx5.cpp

```diff
@@ -30,25 +30,99 @@
+#if defined(HAVE_INFINIBAND_MLX5_HW_H)
 
 #include <sys/mman.h>
+#include "vma/hw/mlx5/wqe.h"
 #include "cq_mgr_mlx5.h"
 #include "vma/util/utils.h"
 #include "vlogger/vlogger.h"
-#define qp_logerr __log_info_err
+#include "ring_simple.h"
+
+#undef  MODULE_NAME
+#define MODULE_NAME 	"qpm_mlx5"
+#define qp_logpanic 	__log_info_panic
+#define qp_logerr	__log_info_err
+#define qp_logwarn	__log_info_warn
+#define qp_loginfo	__log_info_info
+#define qp_logdbg	__log_info_dbg
+#define qp_logfunc	__log_info_func
+#define qp_logfuncall	__log_info_funcall
+
+//
+void qp_mgr_eth_mlx5::init_sq()
+{
+	struct verbs_qp *vqp = (struct verbs_qp *)m_qp;
+	m_hw_qp = (struct mlx5_qp*)container_of(vqp, struct mlx5_qp, verbs_qp);
+	m_qp_num	 = m_hw_qp->ctrl_seg.qp_num;
+	m_sq_wqes	 = (volatile struct mlx5_wqe64 (*)[])(uintptr_t)m_hw_qp->gen_data.sqstart;
+	m_sq_wqe_hot	 = &(*m_sq_wqes)[0];
+	m_sq_wqes_end	 = (uint8_t*)m_hw_qp->gen_data.sqend;
+
+	m_sq_db		 = &m_hw_qp->gen_data.db[MLX5_SND_DBR];
+	m_sq_bf_reg	 = m_hw_qp->gen_data.bf->reg;
+	m_sq_bf_buf_size = m_hw_qp->gen_data.bf->buf_size;
+
+	m_sq_wqe_hot_index = 0;
+	m_sq_bf_offset	 = m_hw_qp->gen_data.bf->offset;
+/*
+ * Preliminary fill WQE for completetion request
+ * TODO: temporary - exclude for dynamic WQE
+ */
+	unsigned int comp = NUM_TX_WRE_TO_SIGNAL_MAX;
+
+	for (uint32_t i = 0; i < m_tx_num_wr; ++i) {
+		volatile struct mlx5_wqe64 *wqe = &(*m_sq_wqes)[i];
+
+		memset((void *)(uintptr_t)wqe, 0, sizeof(struct mlx5_wqe64));
+		wqe->eseg.inline_hdr_sz = htons(MLX5_ETH_INLINE_HEADER_SIZE);
+		wqe->eseg.cs_flags = MLX5_ETH_WQE_L3_CSUM | MLX5_ETH_WQE_L4_CSUM;
+		wqe->ctrl.data[1] = htonl((m_qp_num << 8) | 4);
+		//wqe->dseg.lkey = (m_p_ring->get_lkey());
+		/* Store the completion request in the WQE. */
+		if (--comp == 0) {
+			wqe->ctrl.data[2] = htonl(8);
+			comp = NUM_TX_WRE_TO_SIGNAL_MAX;
+		}
+		else
+			wqe->ctrl.data[2] = 0;
+	}
+	m_sq_wqe_hot->ctrl.data[0] = htonl(MLX5_OPCODE_SEND);
+
+	qp_logdbg("%p: %p allocated for %d QP and configured %u WRs BlueFlame buf_size: %d offset: %d",
+		  this, m_qp, m_qp_num, m_tx_num_wr, m_sq_bf_buf_size, m_sq_bf_offset);
+}
 
 qp_mgr_eth_mlx5::qp_mgr_eth_mlx5(const ring_simple* p_ring, const ib_ctx_handler* p_context, const uint8_t port_num,
 		struct ibv_comp_channel* p_rx_comp_event_channel, const uint32_t tx_num_wr, const uint16_t vlan) throw (vma_error):
-		qp_mgr_eth(p_ring, p_context, port_num, p_rx_comp_event_channel, tx_num_wr, vlan, false) {
+	qp_mgr_eth(p_ring, p_context, port_num, p_rx_comp_event_channel, tx_num_wr, vlan, false)
+	,m_hw_qp(NULL)
+	,m_sq_wqe_idx_to_wrid(NULL)
+	,m_sq_wqes(NULL)
+	,m_sq_wqe_hot(NULL)
+	,m_sq_wqes_end(NULL)
+	,m_sq_db(NULL)
+	,m_sq_bf_reg(NULL)
+	,m_qp_num(0)
+	,m_sq_wqe_hot_index(0)
+	,m_sq_bf_offset(0)
+	,m_sq_bf_buf_size(0)
+	,m_sq_wqe_cntr(0)
+{
 	if(configure(p_rx_comp_event_channel)) {
 		throw_vma_exception("failed creating qp_mgr_eth");
 	}
+
+	init_sq();
```

Comment:
Yep, it was pretty complex and ugly to have two PostSend (new and old VMAPOLLs), it was simplified with enabling VMAPOLL to use new TX PostSend PRM. This was primary reason to do such task now.